src/backend/concurrency/eager_write_txn_manager.cpp

```diff
@@ -0,0 +1,672 @@
+      auto tuple_slot = tuple_entry.first;
+
+      // we don't have reader lock on insert
+      if (tuple_entry.second == RW_TYPE_INSERT ||
+          tuple_entry.second == RW_TYPE_INS_DEL) {
+        continue;
+      }
+
+      RemoveReader(tile_group_header, tuple_slot, txn->GetTransactionId());
+    }
+  }
+  LOG_INFO("release EWreader finish");
+}
+
+// check whether the current transaction owns the tuple.
+// this function is called by update/delete executors.
+bool EagerWriteTxnManager::IsOwner(
+    const storage::TileGroupHeader *const tile_group_header,
+    const oid_t &tuple_id) {
+  auto tuple_txn_id = tile_group_header->GetTransactionId(tuple_id);
+  return tuple_txn_id == current_txn->GetTransactionId();
+}
+
+// if the tuple is not owned by any transaction and is visible to current
+// transaction.
+// this function is called by update/delete executors.
+bool EagerWriteTxnManager::IsOwnable(
+    const storage::TileGroupHeader *const tile_group_header,
+    const oid_t &tuple_id) {
+  auto tuple_txn_id = tile_group_header->GetTransactionId(tuple_id);
+  auto tuple_end_cid = tile_group_header->GetEndCommitId(tuple_id);
+  LOG_INFO("IsOwnable txnid: %lx end_cid: %lx", tuple_txn_id, tuple_end_cid);
+  return tuple_txn_id == INITIAL_TXN_ID && tuple_end_cid == MAX_CID;
+}
+
+bool EagerWriteTxnManager::AcquireOwnership(
+    const storage::TileGroupHeader *const tile_group_header,
+    const oid_t &tile_group_id __attribute__((unused)), const oid_t &tuple_id) {
+  LOG_INFO("AcquireOwnership");
+  assert(IsOwner(tile_group_header, tuple_id) == false);
+
+  // Try to get write lock
+  auto current_tid = current_txn->GetTransactionId();
+  auto old_tid = INITIAL_TXN_ID;
+
+  GetEwReaderLock(tile_group_header, tuple_id);
+
+  old_tid =
+      tile_group_header->SetAtomicTransactionId(tuple_id, old_tid, current_tid);
+
+  // Check if we successfully get the write lock
+  bool res = (old_tid == INITIAL_TXN_ID);
+  if (!res) {
+    LOG_INFO("Fail to acquire write lock. Set txn failure.");
+    ReleaseEwReaderLock(tile_group_header, tuple_id);
+    return false;
+  }
+
+  // Install wait for dependency on all reader txn
+  auto ptr = GetEwReaderList(tile_group_header, tuple_id);
+  while (ptr->next != nullptr) {
+    // try to install a wait for in every reader
+    auto reader_tid = ptr->next->txn_id_;
+    // skip my self
+    if (reader_tid == current_tid) {
+      ptr = ptr->next;
+      continue;
+    }
+
+    // Lock the txn bucket
+    {
+      std::lock_guard<std::mutex> lock(running_txn_map_mutex_);
+      if (running_txn_map_.count(reader_tid) != 0) {
+        // Add myself to the wait_for set of reader
+        auto reader_ctx = running_txn_map_[reader_tid];
+        LOG_INFO("Add dependency to %lu", reader_tid);
+        if (reader_ctx->wait_list_.insert(current_tid).second == true) {
+          // New dependency
+          current_txn_ctx->wait_for_counter_++;
+        }
+      } else {
+        assert(false);
```

Comment:
Could the assert just be `assert(running_txn_map_.count(reader_tid) != 0)` and then remove the if condition?

Or perhaps something else besides asserting should happen when you don't find the tid in the txn map?